### PR TITLE
[Inference] Fix an unsafe fork for frontend replicas and support Nemotron VL vision preprocessing.

### PR DESCRIPTION
**Pull request by cspades**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

- Fix a `fork` that causes a completely underspecified hang in NeMo-RL that only appears at scale.
  - This is the only change: `megatron/core/inference/text_generation_server/dynamic_text_gen_server/text_generation_server.py`
- Support NemotronVL-style image/video preprocessing for per-tubelet headers/footers and frame timestamps in the prompt for the Omni model. This is needed for parity with NeMo-RL training inputs.
  - TL;DR Before we did `<image>....<image></img>` but for that sampling style we need `Frame 1 sampled at 0.00 seconds and frame 2 sampled at 0.99 seconds: <image>....<image></img>\n` _**per-tubelet**_. So we were sending in WAY less tokens into the inference model than we are sending into the training model in RL.
- Fold in a commit (https://github.com/NVIDIA/Megatron-LM/commit/8b380372cf5ae81c56540d1f77958eb2085958fd) that was lost from my previous serialization PR: https://github.com/NVIDIA/Megatron-LM/pull/7352

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by sidsingh-nvidia — APPROVED** [review]
LGTM. Changing to spawn is legit. We had made the same choice for the DP coordinator, in the past.

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py [hunk 1]
```diff
@@ -272,6 +323,25 @@ def preprocess_image_bytes_list(
...
 
+    if config.dynamic_resolution_model_length is not None:
+        model_length = int(config.dynamic_resolution_model_length)
+        if model_length <= 4:
+            raise ValueError("dynamic_resolution_model_length must be greater than 4.")
+        merge_size = max(int(config.spatial_merge_size), 1)
+        model_patch_budget = (model_length - 4) * (merge_size * merge_size)
+        request_patch_budget = max(
+            model_patch_budget, int(config.dynamic_resolution_min_patches) * len(image_bytes_list)
+        )
```
Comment (RPrenger) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:354]: `dynamic_resolution_model_length`: isn't this dividing the wrong way?

```
request_patch_budget = max(
    model_patch_budget, int(config.dynamic_resolution_min_patches) * len(image_bytes_list)
)
```
...
`config = replace(config, dynamic_resolution_max_patches=max(min_patches, request_patch_budge`t))

`request_patch_budget` looks like it's meant to be a total budget for the request (hence multiplying by len(image_bytes_list)), but it gets assigned straight to `dynamic_resolution_max_patches`, which is a per-image cap everywhere else in this file — and the same config is reused unchanged for every image in the loop, so nothing divides it back down by image count.

Concretely, with model_length=5, min_patches=2, configured_max=32, merge_size=2, per-image max_patches goes 4 → 6 → 20 → 32 as image count goes 1 → 3 → 10 → 50, instead of shrinking. That's the opposite of the docstring ("budget used by processors that divide capacity across images"). Shouldn't this divide `request_patch_budget` by `len(image_bytes_list)` before applying it per-image? Or is `dynamic_resolution_max_patches` meant differently here than elsewhere in the file?

Comment (cspades) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:354]: This truly could be a bug with all HF Omni image preprocessors, just take these two as a starting point, neither normalize back to num images to get patches per single frame/image:

- https://huggingface.co/nvidia/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-BF16/blob/main/image_processing.py#L100
- https://huggingface.co/nvidia/NVIDIA-Nemotron-3.5-Super-midtrain-67B-vision-pretrained/blob/main/image_processing_nemotron_h_omni.py#L107

Comment (cspades) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:354]: Maybe we can just implement it correctly, and then see if it matters for vLLM / HF parity or not. We might not need these egregiously large upper bounds.

Edit: Oh, it definitely matters.
> Adding // len(images) would change 2,202 of 15,777 images — about 14% — and every change is a downscale. The per-image cap would fall to 8,190 at N=8, 5,460 at N=12, 4,095 at N=16, which starts clipping the common 1080×1920 frames (8,160 natural patches) from N=9 upward. You'd be sacrificing real resolution on 14% of images to respect a 16,384-token budget that doesn't describe your run.

Comment (cspades) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:354]: Let's just keep it this way for parity, but hopefully it gets fixed in a future PR. I'll add a comment.

```
        # NOTE(@cspades): Oddly, Nemotron VL 3.5 uses more patches per image when given more images.
        # This is counter-intuitive and likely a bug:
        # https://huggingface.co/nvidia/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-BF16/blob/main/image_processing.py#L98
        # Until fixed, manually control patches per image using dynamic_resolution_max_patches.
        if len(image_bytes_list) > 1:
            _warn_multi_image_patch_budget()
```

**Review by RPrenger — APPROVED** [review]
LGTM!

File: megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py [hunk 2]
```diff
@@ -28,6 +29,95 @@
...
+        media_text = prompt_spec.prefix + prompt_spec.model_token + prompt_spec.suffix
+        if prompt_spec.include_frame_timestamps_for_nemotron_vl:
+            media_text = " and ".join(descriptions) + ": " + media_text
+        tubelet_text.append(media_text)
+    return "\n".join(tubelet_text)
+
+
+def _replace_compact_video_slots(
+    tokens: list[list[int]],
+    *,
```
Comment (Phlip79) [inline megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py:68]: Why is there a `*`?

File: megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py [hunk 3]
```diff
@@ -58,7 +148,16 @@ def run_one_forward_step(
         return super().run_one_forward_step(inference_input, recv_buffer_seq_len)
 
     def expand_image_tokens(
-        self, tokens, num_tiles=None, imgs_sizes=None, num_frames=None, *, image_token_id=None
+        self,
+        tokens,
+        num_tiles=None,
+        imgs_sizes=None,
+        num_frames=None,
+        *,
```
Comment (Phlip79) [inline megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py:156]: I know this was already here, but why is there a `*`?

File: megatron/core/inference/config.py [hunk 4]
```diff
@@ -282,6 +282,31 @@ class ImageProcessingConfig:
     max_num_tiles: int = 1
     use_thumbnail: bool = False
     num_img_embeddings_per_tile: int = 0
+    dynamic_resolution_model_length: Optional[int] = None
+    """Model-length budget used by processors that divide capacity across images."""
+    dynamic_resolution_rounding_mode: str = "ceil"
+    """Patch-grid rounding contract: ``ceil`` or ``round_plus_half``."""
+    dynamic_resolution_resize_mode: str = "pil"
+    """Resize contract: ``pil`` or ``torch_bicubic_antialias``."""
```
Comment (Phlip79) [inline megatron/core/inference/config.py]:
```suggestion
    dynamic_resolution_rounding_mode: Literal["ceil", "round_plus_half"] = "ceil"
    """Patch-grid rounding contract: ``ceil`` or ``round_plus_half``."""
    dynamic_resolution_resize_mode: Literal["pil", "torch_bicubic_antialias"] = "pil"
    """Resize contract: ``pil`` or ``torch_bicubic_antialias``."""
```

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py [hunk 5]
```diff
@@ -9,15 +9,31 @@
...
 from megatron.core.models.vision.encoder_registry import REGISTRY as _ENCODER_REGISTRY
 
+logger = logging.getLogger(__name__)
+
+
+@lru_cache(maxsize=1)
+def _warn_multi_image_patch_budget() -> None:
+    """Warn once per process about Nemotron's counter-intuitive image budget."""
+    logger.warning(
+        "Nemotron VL 3.5 may use more patches per image when given more images. "
```
Comment (Phlip79) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:32]: Is this inherently Nemotron-specific?

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py [hunk 6]
```diff
@@ -77,6 +112,7 @@ def dynamic_res_preprocess(
     pixel_shuffle=False,
     spatial_merge_size=1,
     video_maintain_aspect_ratio=None,
+    rounding_mode="ceil",
```
Comment (Phlip79) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py]:
```suggestion
    rounding_mode: Literal["ceil", "round_plus_half"] = "ceil",
```

**Review by Phlip79 — APPROVED** [review]
Might be good to create an enum for `ceil`/`round_plus_half` as its used across files.

File: megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py (hunk 2 above)
Comment (cspades) [reply megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py:68]: AI-gen'd, but this just signifies the end of all positional arguments. Everything after MUST be keyword. It helps prevent people from breaking function signatures by adding more keyword arguments that could be interpreted as keyword or positional without this `*`.

File: megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py (hunk 3 above)
Comment (cspades) [reply megatron/core/inference/model_inference_wrappers/multimodal/nemotron_omni_inference_wrapper.py:156]: see above, unless it irks you I don't think it matters and is good practice to use for arg organization anyway

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py (hunk 5 above)
Comment (cspades) [reply megatron/core/inference/text_generation_server/dynamic_text_gen_server/image_preprocessing.py:32]: Yup, but technically anyone who uses `dynamic_resolution_model_length` will activate this path.

So we should think of it as a bugged version of a `dynamic_resolution_model_length`-drived upper bound on patches / embeddings per image that will hopefully be fixed with the next Nemotron and can be manually set to a specific maximum as a WAR.
